### PR TITLE
Dr. Acula: Untold

### PR DESCRIPTION
**Pull request by Kurfursten**
Reopened due to DMCA. Original spritework by @Blithering . Old: #12506

I think everything from the previous PR was fixed except for updating the UI, but I'm not sure it makes sense to use Nano for only one bot.

:cl:
* rscadd: Added Dr. Acula, the blood donation bot. He's orderable in the 'blood drive' crate at Cargo, but it's expensive. Ask the CMO to cough up from the department account.
* rscadd: There are now three different rewards for donating blood.

### CODE REVIEW

**Comment by jknpj** [comment]
>I'm not sure it makes sense to use Nano for only one bot.

No simple bot boy, no simple bot!

**Comment by PJB3005** [comment]
Just use Nano.

File: code/game/machinery/bots/draculabot.dm [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+/obj/machinery/bot/bloodbot/attack_hand(mob/user as mob)
+	if(..())
+		return
+	var/dat
+	dat += "<TT><B>Acula-class Blood Donation Bot v0.1</B></TT><BR><BR>"
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		var/datum/reagent/blood/B = H.get_blood(H.vessel)
+		if(B.data["virus2"])
+			dat += "<span class='danger'>WARNING: Viral agent detected. Ineligable for blood donation.</span><BR>"
```
Comment (despotate) [inline code/game/machinery/bots/draculabot.dm]: **Ineligible**

File: code/game/machinery/bots/draculabot.dm [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+		dat += "<TT>Blood Storage: "
+		var/counter = 0
+		if(!contents.len)
+			dat += "There are no blood packs available."
+		for (var/obj/item/weapon/reagent_containers/blood/E in contents)
+			counter++
+			dat += "Slot [counter]: [E] <A href='?src=\ref[src];slot=\ref[E]'>(Eject)</A><BR>"
+		dat += "The speaker switch is [src.quiet ? "off" : "on"]. <a href='?src=\ref[src];togglevoice=[1]'>Toggle</a><br>"
+
+	user << browse("<HEAD><TITLE>Dr. Acula Controls</TITLE></HEAD>[dat]", "window=autoblood")
```
Comment (PJB3005) [inline code/game/machinery/bots/draculabot.dm]: If you're not gonna use Nano, at least make it use browser datums.

File: code/game/machinery/bots/draculabot.dm [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+		playsound(get_turf(src), 'sound/effects/splat.ogg', 50, 1)
+		spawn(1 SECONDS)
+			drink(H)
+
+/obj/machinery/bot/bloodbot/proc/dispense_reward()
+	speak(pick("Enjoy!","Come again!","Donate often!"))
+	var/path = pick(possible_rewards)
+	new path(get_turf(src))
+
+/obj/machinery/bot/bloodbot/proc/speak(var/message)
```
Comment (PJB3005) [inline code/game/machinery/bots/draculabot.dm:266]: Wouldn't it be better to use `say` than `visible_message`?

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: code/game/machinery/bots/draculabot.dm (hunk 3 above)
Comment (Kurfursten) [reply code/game/machinery/bots/draculabot.dm:266]: Sure, I don't have a strong preference.

**Review by Probe1 — APPROVED** [review]
I only skimmed the code but the design seems solid.

**Comment by Probe1** [comment]
coderkillerwip2

Waiting on changes requested to be resolved.

**Comment by Kurfursten** [comment]
A spaceman's greatest weapon

is patience

**Comment by jknpj** [comment]
>A spaceman's greatest weapon is

autism

**Comment by Probe1** [comment]
A coders greatest threat is CONFLICTU and I'd like to see what you've worked so hard on merged.

**Comment by Kurfursten** [comment]
It may look like it's near ready to go, but it's not. If it went to live like this, it would still be the half-assed version that only accepts empty bags and this really deserves to be full-assed. Rewriting how that works is going to take a little time.

**Comment by Probe1** [comment]
That attitude is why you are an awesome coder.

**Comment by Probe1** [comment]
THEN

HE

DIED

**Comment by Kurfursten** [comment]
I'm still getting the mysterious map bug, can't code until we figure out what's up with that.

**Comment by Probe1** [comment]
We've resolved that issue here.  https://github.com/d3athrow/vgstation13/commit/3275888e64aec79e673170e76588925ce5a4eb4b#commitcomment-20299318

**Comment by Probe1** [comment]
THEN
HE
DIED
AGAIN

**Comment by Blithering** [comment]
We must perform unholy rituals to raise this PR from it's coffin

**Comment by Probe1** [comment]
Reopen if you plan on working on it.

**Comment by Kurfursten** [comment]
You can't reopen branches closed by collaborators

**Comment by Exxion** [comment]
You can ask a collab to do it

**Comment by Iamgoofball** [comment]
Also you totally can I do it all the time

**Comment by Exxion** [comment]
False.

**Comment by Kurfursten** [comment]
Curious is the coder's art. His PR unopenable by his own hand.

**Comment by Probe1** [comment]
Seize this momentum! Push on to the task's end!

**Comment by Kurfursten** [comment]
Status update
* Normal (nonemagged) is 100% tested and working
* Emagged mode still needs some bugfixing
* Needs browser datum or NanoUI
* Need to be able to get it from somewhere or map it in

But I haven't fucked with UIs in at least a year. I can do browser datums since they're easy, but I'm willing to do Nano if PJB will link the documentation or whatever.

**Comment by PJB3005** [comment]
Just go browser if nano is too much of a hassle.

**Comment by Kurfursten** [comment]
I don't mind writing it in Nano if you'll explain how to do it or link me to documentation that shows how.

**Comment by Kurfursten** [comment]
At some point in the line did you lose your zeal for Nano?

**Comment by PJB3005** [comment]
https://wiki.baystation12.net/NanoUI

**Comment by Kurfursten** [comment]
This is ready to merge

I went with a datum instead of nano because templ files were complicated

File: code/game/machinery/bots/draculabot.dm [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+		for (var/obj/item/weapon/reagent_containers/blood/E in contained_bags)
+			counter++
+			dat += "Slot [counter]: [E] ([100 * E.reagents.total_volume / E.reagents.maximum_volume]% filled)<A href='?src=\ref[src];slot=\ref[E]'>(Eject)</A><BR>"
+		dat += "</TT>The speaker switch is [src.quiet ? "off" : "on"]. <a href='?src=\ref[src];togglevoice=[1]'>Toggle</a><br>"
+
+	dat = jointext(dat,"")
+	var/datum/browser/popup = new(usr, "bloodbot", "[name]", 575, 400, src)
+	popup.set_content(dat)
+	popup.open()
+	onclose(user, "bloodbot")
```
Comment (PJB3005) [inline code/game/machinery/bots/draculabot.dm]: This is already done by browser datums.

File: code/game/machinery/bots/draculabot.dm [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+		var/counter = 0
+		if(!contents.len)
+			dat += "There are no blood packs available.<BR>"
+		for (var/obj/item/weapon/reagent_containers/blood/E in contained_bags)
+			counter++
+			dat += "Slot [counter]: [E] ([100 * E.reagents.total_volume / E.reagents.maximum_volume]% filled)<A href='?src=\ref[src];slot=\ref[E]'>(Eject)</A><BR>"
+		dat += "</TT>The speaker switch is [src.quiet ? "off" : "on"]. <a href='?src=\ref[src];togglevoice=[1]'>Toggle</a><br>"
+
+	dat = jointext(dat,"")
+	var/datum/browser/popup = new(usr, "bloodbot", "[name]", 575, 400, src)
```
Comment (PJB3005) [inline code/game/machinery/bots/draculabot.dm]: Make the UI key (`"bloodbot"`) be `\ref[src]` instead. No issues having two UIs open at once!

File: code/game/machinery/bots/draculabot.dm [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+	update_icon()
+
+/obj/machinery/bot/bloodbot/attack_paw(mob/user as mob)
+	return attack_hand(user)
+
+/obj/machinery/bot/bloodbot/attack_hand(mob/user as mob)
+	if(..())
+		return
+	var/dat
+	dat += "<TT><B>Acula-class Blood Donation Bot v1.0</B></TT><BR><BR>"
```
Comment (PJB3005) [inline code/game/machinery/bots/draculabot.dm:70]: Don't browser datums have a title header already? Use that instead.

File: code/game/machinery/bots/draculabot.dm [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		var/datum/reagent/blood/B = H.get_blood(H.vessel)
+		if(B.data["virus2"])
+			dat += "<span class='danger'>WARNING: Viral agent detected. Ineligible for blood donation.</span><BR>"
```
Comment (PJB3005) [inline code/game/machinery/bots/draculabot.dm]: Are these styles even included in the browser datum CSS?

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: code/game/machinery/bots/draculabot.dm (hunk 6 above)
Comment (Kurfursten) [reply code/game/machinery/bots/draculabot.dm:70]: This makes the font impact or whatever, it's a sarif font

I use it for the bloodbags down below too, just to make the text look different

**Comment by Kurfursten** [comment]
Okay, now you can merge it

**Comment by Probe1** [comment]
Ping @PJB3005 to check his review.

**Review by PJB3005 — APPROVED** [review]

**Comment by Probe1** [comment]
At this point I don't even know what this PR does anymore but the time is now.

**Comment by Kurfursten** [comment]
It adds a crate to cargo for a fuckton of money that has a bot

The bot comes with 8 preloaded empty bags, and also 4 free cookies, so it's not a small deal. The bot will then collect blood donations from crew members who volunteer and have at least 90% of their blood. Every time 55u are donated, a new reward is dispensed.

Doctors still need to come by to collect and replace bags though.

**Comment by Probe1** [comment]
(I was just kidding with you I'm not that bad)

**Comment by SonixApache** [comment]
yaaaay

**Comment by Blithering** [comment]
These nightmarish PJB reviews can be felled, they can be beaten!

**Comment by Probe1** [comment]
In time you will know the tragic extent of my compile failings.

**Comment by Kurfursten** [comment]
Glittering merges, feedback and 👍 -- paid for in pomfcoins.

**Pull request merged**